### ISSUE TITLE
SetState does not handle well nested SetState calls

### ISSUE THREAD

**Issue by amieres** [facebook/react#8132]
This could be a bug (or a nice to have feature) :
SetState does not handle well nested SetState calls

The successive calls are put in the _pendingStateQueue but not resolved, because the queue reference is cleared too soon.
My suggestion would be to change function **_processPendingState** like this:

```
  _processPendingState: function (props, context) {
    var inst = this._instance;
    var queue = this._pendingStateQueue;
    var replace = this._pendingReplaceState;
    this._pendingReplaceState = false;
//    this._pendingStateQueue = null; //////  <<<<<<<<<<<<<
    if (!queue) {
      return inst.state;
    }

    if (replace && queue.length === 1) {
      this._pendingStateQueue = null; ////// <<<<<<<<<<<<<
      return queue[0];
    }

    var nextState = _assign({}, replace ? queue[0] : inst.state);
    for (var i = replace ? 1 : 0; i < queue.length; i++) {
      var partial = queue[i];
      _assign(nextState, typeof partial === 'function' ? partial.call(inst, nextState, props, context) : partial);
    }
    this._pendingStateQueue = null;  ////// <<<<<<<<<<<<<

    return nextState;
  },
```

**Comment by gaearon** [comment]
Can you please show the code you're having issues with?

**Comment by amieres** [comment]
Unfortunately I'm coding in F# transpiled into javascript which makes it almost unreadable.

I'm dealing with messages that change the state and some times the calls are recursive and they usually initiate from a change of state.
Something like this:

```
setNewValue: function(v) {
   setState(function(state) {
       state.newValue = v;
       showMessage("changing value to " + v + "...");
   });
}

showMessage: function(msg) {
   setState(function(state) {
      state.message = msg;
   });
}

```

Because the call to showMessage happens within the first call to setState, it does not get added to the current _pendingStateQueue because the reference has already been cleared. Instead it gets added to a new queue that is not executed until later when a future call to setState happens.

My suggested change simply maintains the reference to the queue until just before returning which allows for new setState to be added to the queue and be processed.

**Comment by milesj** [comment]
Anyway you can change your F# code to trigger the message in the `setState` callback like so?

```
setNewValue: function(v) {
   setState(function(state) {
       state.newValue = v;
   }, function() {
       showMessage("changing value to " + v + "...");
   });
}

showMessage: function(msg) {
   setState(function(state) {
      state.message = msg;
   });
}
```

**Comment by amieres** [comment]
Thanks for the suggestion Miles.
That would not work for me.
The code I posted is just a small sample. 
There are more calls happening as the controls communicate among them.
That would trigger multiple renders each time.

The other problem is that the code would get separated.
I am using what is called the elm architecture: https://guide.elm-lang.org/architecture/
which lines up very well with React because there is a State (**model**),  a  Render function (**view**) and one **update** function where the state gets changed.
That function acts as a messaging processor and thus it is recursive at times.

In fact prior to the current implementation I was using a mailbox agent for the update function and it worked well with React with one exception, the textbox input. Because the mailbox is asynchronous any change to an input would make the cursor jump to the end: (https://github.com/facebook/react/issues/955).

So I decided to remove the mailbox to make it synchronous. Now textbox works but recursive calls do not work.

Of course I made a local copy of react.js and with the above mentioned change it works for me.

**Comment by gaearon** [comment]
First of all, this is not correct usage:

```js
   setState(function(state) {
       state.newValue = v;
       showMessage("changing value to " + v + "...");
   });
```

State should be treated as immutable, and you must never assign to it. Instead return the update:

```js
   setState(function(state) {
       return { newValue: v };
   });
```

Since you're not using previous state, you might as well pass an object right away:

```js
   setState({ newValue: v });
```

Secondly, calling `setState()` from inside an updater function is not supported. (Maybe we should throw in this case?

The updater function is supposed to be pure and return the next state. It should not execute any side effects.

**Comment by amieres** [comment]
Dan, 
Thanks for your help.

"First of all, this is not correct usage:`"

You are right. I meant to use return not assign, my mistake.
I was paraphrasing from code in F# and got it wrong. 
It should read something like this:

```
setNewValue: function(v) {
   setState(function(state) {
       showMessage("changing value from " + state.newValue + " to "+ v + "...");
       return { newValue: v };
   });
}

showMessage: function(msg) {
   setState(function(state) {
      return state.messages.push(msg);
   });
}

```
"Secondly, calling setState() from inside an updater function is not supported. (Maybe we should throw in this case?"

I imagined that could be the case, that is why I did not call it a bug, but a possible feature.
It is your decision if you want to allow recursive calls or not. 
It does not seem too difficult to allow it with the modification posted above.
It would make React more versatile.


OTOH, throwing will be harder (I think)

Cheers
